File: services/ReadingService1/src/main/java/com/se1853_jv/reading/model/ReadingWorkflowHighlight.java
package com.se1853_jv.reading.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.Column;
import jakarta.persistence.EmbeddedId;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinColumns;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.PrePersist;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDateTime;

@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "reading_workflow_highlight")
public class ReadingWorkflowHighlight {

    @EmbeddedId
    private ReadingWorkflowHighlightId id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumns({
            @JoinColumn(name = "collection_id", referencedColumnName = "collection_id", insertable = false, updatable = false),
            @JoinColumn(name = "paper_id", referencedColumnName = "paper_id", insertable = false, updatable = false),
            @JoinColumn(name = "user_id", referencedColumnName = "user_id", insertable = false, updatable = false)
    })
    @JsonIgnore
    private ReadingWorkflow readingWorkflow;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "highlight_id", referencedColumnName = "id", insertable = false, updatable = false)
    private Highlight highlight;

    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    @PrePersist
    public void prePersist() {
        this.createdAt = LocalDateTime.now();
    }
}

File: services/paper-service/src/main/kotlin/com/se1853_jv/model/ReadingWorkflowHighlight.kt
package com.se1853_jv.model

import jakarta.persistence.*
import java.io.Serializable

/**
 * Junction table: ReadingWorkflow_Highlight
 * Link Highlight với ReadingWorkflow (many-to-many relationship)
 */
@Entity
@Table(name = "ReadingWorkflow_Highlight")
@IdClass(ReadingWorkflowHighlightId::class)
data class ReadingWorkflowHighlight(
    @Id
    @Column(name = "RWCollectionId", length = 36, nullable = false)
    val rwCollectionId: String,

    @Id
    @Column(name = "RWPaperId", length = 36, nullable = false)
    val rwPaperId: String,

    @Id
    @Column(name = "RWUsersid", length = 36, nullable = false)
    val rwUserId: String,

    @Id
    @Column(name = "Highlightid", length = 36, nullable = false)
    val highlightId: String,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumns(
        JoinColumn(name = "RWCollectionId", referencedColumnName = "CollectionId", insertable = false, updatable = false),
        JoinColumn(name = "RWPaperId", referencedColumnName = "PaperId", insertable = false, updatable = false),
        JoinColumn(name = "RWUsersid", referencedColumnName = "UsersId", insertable = false, updatable = false)
    )
    val readingWorkflow: ReadingWorkflow? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "Highlightid", referencedColumnName = "id", insertable = false, updatable = false)
    val highlight: Highlight? = null
)

data class ReadingWorkflowHighlightId(
    val rwCollectionId: String,
    val rwPaperId: String,
    val rwUserId: String,
    val highlightId: String
) : Serializable







